### PR TITLE
Added Backtracking algorithm (N-Queen Problem) [Python]

### LINKED ISSUE #265: Backtracking algorithm (N-Queen Problem) [Python]
#### This issue is about [Backtracking algorithm (N-Queen Problem)](https://en.wikipedia.org/wiki/Eight_queens_puzzle).

- [X] Issue Title is in form `Algo/DS Name [Language]`.
- [X] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [X] This Algo/DS is worth adding to this repository.
- [X] Assign this issue to me (I want to work on this).
- [X] I will send my PR only after I'm assigned to this issue by any maintainer.

A general solution of the N-Queen problem (any board of n X n) in python with backtracking.

### PR DESCRIPTION
**Pull request by sirfoga**
Fixes #265 

#### By submitting this pull request I confirm I've read and complied with the below declarations.

- [X] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [X] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [X] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [X] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [X] This pull request will be closed if I fail to update it even once in a continuous time span of 7 days.

### CODE REVIEW

File: n_queen_problem/n_queen_problem.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        for column in range(1, board_size + 1):
+            if not under_attack(column, solution):
+                solutions.append(solution + [(queens, column)])
+
+    return solutions
+
+
+def main():
+    board_size = 10
+    solutions = solve(board_size, board_size)
```
Comment (Monal5031) [inline n_queen_problem/n_queen_problem.py]: Can there be more than `n` queens for a `n X n` board?

Comment (sirfoga) [inline n_queen_problem/n_queen_problem.py]: This implementation is more general .. given a 8 x 8 board you could want to find the solutions with 7 queens .. should I revert to the original version?

Comment (Monal5031) [inline n_queen_problem/n_queen_problem.py]: No its fine, nice thought though 👍

**Comment by Monal5031** [comment]
@sirfoga Please remove `{}` from the PR description as they won't let the issue be linked with this PR.

File: n_queen_problem/n_queen_problem.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    if queens < 1:  # trivial cases
+        return [[]]
+
+    if queens > board_size:
+        return []  # no solutions
+
+    return get_solutions(board_size, queens)
+
+
+def n_queen_problem(board_size):
```
Comment (Monal5031) [inline n_queen_problem/n_queen_problem.py]: The previous version was better as we had a flexibility to choose the number of queens. Please revert this change.

**Comment by aviaryan** [comment]
Fix codacy issues and resolve conflicts.

**Comment by sirfoga** [comment]
@aviaryan the conflict is about the algorithm entry in README ..

File: README.md [hunk 3]
```diff
@@ -91,4 +91,4 @@ Idea by [@Monal5031](https://github.com/Monal5031)
 * [Avi Aryan](https://github.com/aviaryan)
 * [Pratyush Singh](https://github.com/singhpratyush)
 
-Only project maintainers should merge a PR. Other members can add their reviews to a PR but the merging should be done by only a project maintainer.
+Only project maintainers should merge a PR. Other members can add their reviews to a PR but the merging should be done by only a project maintainer.
```
Comment (singhpratyush) [inline README.md]: Why this diff?

File: n_queen_problem/n_queen_problem.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        Size of board to solve
+    :param queens: int
+        Number of queens to place in board
+    :return: list of list of tuple of int
+        Each list contains the coordinates of the queens
+        that solve the problem
+    """
+
+    if queens < 1:  # trivial cases
+        return [[]]
```
Comment (singhpratyush) [inline n_queen_problem/n_queen_problem.py]: Please add a comment explaining this.

File: README.md (hunk 3 above)
Comment (sirfoga) [inline README.md]: @singhpratyush  screenshot from 2017-06-13 11-21-07
I just updated my README with your `master/README.md`

Comment (sirfoga) [inline README.md]: @singhpratyush  screenshot from 2017-06-13 11-21-07
I just updated my README with your `master/README.md`

File: n_queen_problem/n_queen_problem.py (hunk 4 above)
Comment (sirfoga) [inline n_queen_problem/n_queen_problem.py]:
```python
    if queens < 1:  # there are no queens to place
        return [[]]  # 1 solution only: not place any queen

    if queens > board_size:  # more queens than the board can contain
        return []  # pigeonhole principle -> 0 solutions
```

@singhpratyush is it clear enough?

File: README.md (hunk 3 above)
Comment (singhpratyush) [inline README.md]: It should not be there in any case. Please take a look into it.

**Comment by sirfoga** [comment]
@singhpratyush cleande my local repo, now there should only be the entry in README edit and the algorithm addition
